crates/kubelet/src/node.rs

```diff
@@ -58,11 +60,57 @@ macro_rules! retry {
 /// A node comes with a lease, and we maintain the lease to tell Kubernetes that the
 /// node remains alive and functional. Note that this will not work in
 /// versions of Kubernetes prior to 1.14.
-pub async fn create_node(client: &kube::Client, config: &Config, arch: &str) {
-    let node_client: Api<Node> = Api::all(client.clone());
-    let node = node_definition(config, arch);
-    let node =
-        serde_json::from_value(node).expect("failed to deserialize node from node definition JSON");
+pub async fn create_node<P: 'static + Provider + Sync + Send>(
+    client: &kube::Client,
+    config: &Config,
+    provider: Arc<P>,
```

Comment:
this part is a litttle concerning to me. This puts a hard dependency on Providers in the `kubelet` crate. I believe @thomastaylor312 designed this crate to be non-krustlet specific. That way other libraries could consume this to create their own kubelets without the provider system.

Is there a way we can somehow split this out? I like the builder pattern used here. I'm just not sure if this is the right way to allow providers to inject their own information into the node definition.

What if there were two functions:

- `new` returns a NodeBuilder with all the defaults set
- `create` which makes the request to create the node (lines 115+)

What do you think about that approach?

Reply:
I don't think I necessarily have a problem tying this to a `Provider`, but I think @bacongobbler has a point here. It would be better to just pass in a kube client and create the node from there instead of tying the implementation to a `Provider`. As far as I understand, creating a node should be the same for any `Provider` and it should be handled purely by logic external to any specific `Provider`

Reply:
Hmm, perhaps I don't understand the distinction between `krustlet` and the `kubelet` crate. It seems to me that the Provider trait isn't specific to krustlet (I'm assuming you mean the concrete implementations in this repo), but is part of the API for implementing a kubelet, and is defined in the `kubelet` crate.  Provider is already a dependency of the method that calls `create_node`, is it important that `create_node` specifically can be used without a Provider implementation? 

I'm not sure I follow your last sentence, where would providers call `new` and `create`? The goal of this change is to give providers a hook before node registration which does not currently exist. I'm not sure how to add that hook without a dependency on Provider. 

I would support splitting this function into two parts for ergonomics though.

Reply:
@thomastaylor312 I agree that logic for contacting the kube-apiserver is likely the same for each provider but I think construction of the node definition (especially taints) is going to be provider specific. I'm happy to split the function up.

Reply:
I think the main concern for me is it is a little weird to have the `Provider` with a `node` function that configures the builder as part of the trait definition. I think it would be better to use `T::ARCH` and pass that in as an argument to the builder itself. It just seems like an odd step to pass your builder to a provider rather than just getting the information you need from the provider

Reply:
Hmmm, I see. I think the advantage of this is that Providers can completely introspect and even replace the Node object, using the common k8s_openapi type, or not implement the method at all. Using associated constants has some issues I think. It does not allow any dynamic content to be loaded into the node config from the provider (for instance querying the runtime for its version). It also requires `kubelet` to explicitly define what can be customized and limits providers to those fields, and possible results in a very fat trait.

Would it make more sense to have Node return a struct defined in `kubelet` which has fields for a limited set of overrides? That would at least avoid passing in the mutable reference.

Reply:
Let's give this current design a whirl. If we don't like it, we don't have any API stability guarantees yet, so we can remove/refactor down the road. 

@bacongobbler you agree?

Reply:
I'm okay with the current approach for now. If we hear design concerns/feedback from others we can circle back on this.